### PR TITLE
Add option to distribute volume according to geometric distribution

### LINKED ISSUE #315: Volume distribution uneven with high "flow path" density
The original pore volume is distributed evenly over all FlowNet tubes. This can lead to spatially uneven spread of the volume in cases where many flow path are concentrated in a small area as a results of many well connections (for example).

How can we improve this?

### LINKED ISSUE #376: FOIP differences
When the FlowNet model is initialized, even with equal PORV, the FOIP sometimes has significant differences for the Norne model. Why?

### PR DESCRIPTION
**Pull request by wouterjdb**
This PR introduces the option to distribute the FlowNet model volume based on the geometrical distribution in the original model.

---

### Contributor checklist

- [x] :tada: This PR closes #315 and closes #376.
- [x] :scroll: I have broken down my PR into the following tasks:
   - [x] Implement volume distribution method
   - [ ] ~~Add option to choose either model volume or flownet bounding box volume (see remark @wouterjdb)~~ (related issue #368)
   - [ ] ~~Add option to distribute volume per grid cell (see remark from @olwijn)~~ (new issue #358)
   - [x] Expose distribution method to user
   - [x] Test
   - [x] Fix removal of volume by, for example, non-reservoir tube removal. @LonnekevB Could you try to incorporate this?
   - [x] Test again
   - [x] Add error when users ask for a method that depends on a sim model that is not supplied.
- [x] :robot: I have added tests, or extended existing tests, to cover any new features or bugs fixed in this PR.
- [x] :book: I have considered adding a new entry in `CHANGELOG.md`.
- [ ] :books: I have considered updating the documentation.

### CODE REVIEW

**Comment by wouterjdb** [comment]
Comparison of the distribution in the original and new distribution method:

image

Some observations:

- The new method shows a high-volume-ring around the Brugge case. This is correct: the original Brugge model has a large aquifer that is modeled with active gridcells. This volume is now assigned to the nearest modelled FlowNet tube.... 
  
   ❗ **However,** as our bulk volume estimation is coming from the hull between drilled wells I am actually assigning to little volume to the HC-bearing formation in this case. Either I can:
  * use the total volume from the original model OR 
  * only distribute the volume of those cells of the original model that lie inside the hull that was used for the original volume estimation.
  
  What do you think @edubarrosTNO @olelod @olwijn @tayloris?

- The internal tubes in the model have a relatively low volume. This is what one would expect: in areas with a high density of tubes the volume assigned to an individual tube should be lower.

**Comment by olwijn** [comment]
Could this approach to volume distribution perhaps be taken one step further and be used to distribute the flow tube volume over the grid cells of each flow tube? If so, volumes of flow tube cells that connect to or are close to wells would get low volume (since there are many cells from other flow tubes nearby) while cells that arefar away from wells would receive a larger volume.

**Comment by wouterjdb** [comment]
> Could this approach to volume distribution perhaps be taken one step further and be used to distribute the flow tube volume over the grid cells of each flow tube? If so, volumes of flow tube cells that connect to or are close to wells would get low volume (since there are many cells from other flow tubes nearby) while cells that arefar away from wells would receive a larger volume.

That is an interesting thought 👍 - and yes, I think it could be done.

**Comment by wouterjdb** [comment]
@olwijn I tried simply expanding the algorithm to a per cell setting. This works in some areas as expected (see increasing volume further away from the well node):

image

However, you already see some problems in the back. Because of the resolution of the original model not all FlowNet gridcells now get assigned some volume, leading to many holes in the model. Also, at the flownet well connection only one (or none, if unlucky) cell get volume assigned - so the wells are now poorly connected.

image

I think this is a really nice idea, but it requires more work to implement it. Therefore I suggest that I do not implement it here and now but we make an issue out of it to be taken at a later time (could be soon - be at least separate).

**Comment by olwijn** [comment]
an alternative could perhaps be to first assign volumes to flow tubes, as you have already done, and then to somehow distribute that volume over the flow tube cells based on the distance to the end points?

**Comment by wouterjdb** [comment]
> an alternative could perhaps be to first assign volumes to flow tubes, as you have already done, and then to somehow distribute that volume over the flow tube cells based on the distance to the end points?

Let's continue this discussion in #358.

**Review by olelod — CHANGES_REQUESTED** [review]
Nice work. 

Just reminding you of the testing in the config_parser we discussed yesterday, and some other cosmetic changes.

The bulk volume of the FlowNet when using the Voronoi option will always be equal to the simulation model volume, right? There is no need to test or assure that? Or maybe print volumes to screen for the user to see?

File: src/flownet/data/from_flow.py [hunk 1]
```diff
@@ -382,6 +384,63 @@ def get_well_connections(self, perforation_handling_strategy: str) -> pd.DataFra
...
+        volumes to the nearest FlowNet tube midpoint.
+
+        Args:
+            network: FlowNet network instance.
+
+        Returns:
+            A list with multipliers for the total bulk volume for each flownet tube cell.
+
+        Raise:
+            NotImplementedError in case a users asks for this type of distribution but is
```
[the remark this answers is not in the document]
Comment (wouterjdb) [reply src/flownet/data/from_flow.py:402]: I remove the entire I removed the entire Raise-part - it was some older version of the code that did that. Not anymore.

File: src/flownet/data/from_flow.py [hunk 2]
```diff
@@ -382,6 +384,63 @@ def get_well_connections(self, perforation_handling_strategy: str) -> pd.DataFra
...
+            [cell.coordinate for cell in self._grid.cells()]
+        )
+        model_cell_active_state = [cell.active for cell in self._grid.cells()]
+        model_cell_volume = [cell.volume for cell in self._grid.cells()]
+
+        # Determine nearest flow tube for each cell in the original model
+        tree = KDTree(flownet_tube_midpoints)
+        _, matched_indices = tree.query(model_cell_mid_points, k=[1])
+
+        # Distribute the original model's volume per cell, that lies within the FlowNet model's
```
[the remark this answers is not in the document]
Comment (wouterjdb) [reply src/flownet/data/from_flow.py:422]: Done.

File: src/flownet/network_model/_network_model.py [hunk 3]
```diff
@@ -97,6 +99,46 @@ def get_connection_midpoints(self, i: Optional[int] = None) -> np.ndarray:
...
+        """Volume distribution multipliers for each grid cell in the flownet.
+
+        Returns:
+            An np.ndarray with multiplier for each FlowNet grid cell's volume.
+        """
+        return self._initial_cell_volumes
+
+    @initial_cell_volumes.setter
+    def initial_cell_volumes(self, initial_cell_volumes: np.ndarray):
+        """Set
```
[the remark this answers is not in the document]
Comment (wouterjdb) [reply src/flownet/network_model/_network_model.py:113]: Fixed.

File: src/flownet/ahm/_run_ahm.py [hunk 4]
```diff
@@ -442,6 +442,22 @@ def run_flownet_history_matching(
...
 
+    if config.flownet.prior_volume_distribution == "voronoi_per_tube":
+        volumes_per_cell = (
+            field_data.bulk_volume_per_flownet_cell_based_on_voronoi_of_input_model(
+                network,
+            )
+        )
+    elif config.flownet.prior_volume_distribution == "tube_length":
+        volumes_per_cell = network.bulk_volume_per_flownet_cell_based_on_tube_length()
+    else:
```
[the remark this answers is not in the document]
Comment (wouterjdb) [reply src/flownet/ahm/_run_ahm.py:453]: I agree - but I domn't really want to hardcode that `else => tube_length`. So I'll just leave it like this but will write the test in the config parser.

File: src/flownet/config_parser/_config_parser.py [hunk 5]
```diff
@@ -426,6 +426,25 @@ def _to_abs_path(path: Optional[str]) -> str:
                         MK.Description: "Minimum allowed permeability in mD before a tube is removed "
                         "(i.e., its cells are made inactive).",
                     },
+                    "prior_volume_distribution": {
```
[the remark this answers is not in the document]
Comment (wouterjdb) [reply src/flownet/config_parser/_config_parser.py:429]: Done.

**Review by olelod — APPROVED** [review]

**Comment by olelod** [comment]
LGTM!

**Comment by wouterjdb** [comment]
I tested the algorithm but I'm getting lower volumes then I would like to see:

image

and the (low res) FlowNet that has distributed that volume (bulk volume multiplier range 0.9-1.1):

image

So I'm only getting ~75% of the original volume. I think this is a result of FlowNet tubes being remove when they, for example, go through non-reservoir. To be able to keep the volume the same from the original to the new model the algorithm should not distribute volume to tubes that later on deactivate (overwritten with PORV = 0).

I'd guess that the low resolution of the FlowNet exaggerates this phenomenon - but we should consider how to handle this both in this prior volume distribution method as well as the tube-length based distribution method.

**Comment by edubarrosTNO** [comment]
To add to this discussion, here are some thoughts that @olwijn and I had today:
Even if we get a good solution to distribute the pore volume within the FlowNet tubes, we might still have issues in matching the volume of oil-water-gas initially in place (i.e. we have been observing FOIP volumes pretty much off in our current results).

A possible explanation relates to the fact that we rely on initialization by equilibration in the reservoir simulator. Just like we are distributing PORV across the cells / tubes of FlowNet, we must check if the distribution of fluid volumes across tubes is also done properly, otherwise we should control the initialization of saturations within FlowNet and not using the `EQUIL` keyword in OPM-Flow / Eclipse.

**Comment by edubarrosTNO** [comment]
PS: the comment above is based on the fact that my tests with the code of this PR branch did not show to have any impact on the offset of FOIP volumes that we have been observing in the Norne example

**Comment by olelod** [comment]
> So I'm only getting ~75% of the original volume. I think this is a result of FlowNet tubes being remove when they, for example, go through non-reservoir. To be able to keep the volume the same from the original to the flownet model the algorithm should not distribute volume to tubes that later on are deactivate (overwritten with PORV = 0).

The non-reservoir part should not be a problem, I think, since we drop those start/end points before we generate the FlowNet network. The same goes for angles. So I guess the only removal/deactivation happening after generation of the FlowNet is in the rendering of the realizations (where `permx < min_permeability` leads to `PORV = 0`). You see the same thing if you run without defining `min_permeability`?

**Comment by wouterjdb** [comment]
During the assignment of volume to tubes, volume is also assigned to the inactive grid cells. These cells should not be assigned volume.

Potential problem areas:
https://github.com/wouterjdb/flownet/blob/f87804345ce8531055405591115bc2045b7bb3a4/src/flownet/data/from_flow.py#L420

**Comment by wouterjdb** [comment]
> During the assignment of volume to tubes, volume is also assigned to the inactive grid cells. These cells should not be assigned volume.
> 
> Potential problem areas:
> https://github.com/wouterjdb/flownet/blob/f87804345ce8531055405591115bc2045b7bb3a4/src/flownet/data/from_flow.py#L420

This is now fixed. Also a correction for NTG has been added.

**Comment by wouterjdb** [comment]
Tested again:

image

PORV now nearly identical when the bulk volume multiplier is set to 1. The mismatch is now only due to differences in PORO; that is as expected.

File: src/flownet/network_model/_network_model.py [hunk 6]
```diff
@@ -467,6 +513,23 @@ def active_mask(self, model_index) -> List[bool]:
...
+        of the active cells in each tube in the FlowNet and the convex hull on
+        the FlowNet network.
+
+        Args:
+            network: FlowNet network instance.
+
+        Returns:
+            A list with bulk volumes for each flownet tube cell.
+        """
+        return (
```
Comment (olelod) [inline src/flownet/network_model/_network_model.py:527]: Here we don't set the volume in the inactive cells to 0. Should we do it here as well?

Comment (wouterjdb) [reply src/flownet/network_model/_network_model.py:527]: Now the volume is multiplied with the ACTNUM value (0 or 1), which means that the inactive gridcells are set to volume == zero.

**Review by olelod — APPROVED** [review]
LGTM 🚀

**Pull request merged**